{
    "source": "BDSA",
    "name": "BDSA-2021-1271",
    "title": "Django Vulnerable to HTTP Header Injection via URLValidator",
    "description": "Django-based applications are vulnerable to HTTP header injection since `URLValidator` doesn't prohibit newlines and tabs. Django itself isn't vulnerable since `HttpResponse` prohibits newlines in HTTP headers.    \n\nThis issue was introduced in Python **3.9.5** by the fix for [this](https://bugs.python.org/issue43882) issue.",
    "technicalDescription": "Applications which used values with newlines in HTTP response could be exposed to header injection attacks.\n\nThe `URLField` form field which uses `URLValidator` silently removes newlines and tabs on Python **3.9.5** and greater so newlines could only enter  your data if this validator is used outside these form fields.",
    "publishedDate": "2021-05-07T15:35:42.945Z",
    "updatedDate": "2021-05-07T15:35:42.939Z",
    "disclosureDate": "2021-05-06T00:00:00.000Z",
    "solution": "Fixed in:\n* [**2.2.22**](https://www.djangoproject.com/weblog/2021/may/06/security-releases/) by [this](https://github.com/django/django/commit/d9594c4ea57b6309d93879805302cec9ae9f23ff) commit. \n* [**3.1.10**)](https://www.djangoproject.com/weblog/2021/may/06/security-releases/) by [this](https://github.com/django/django/commit/afb23f5929944a407e4990edef1c7806a94c9879) commit. \n* [**3.2.2**](https://www.djangoproject.com/weblog/2021/may/06/security-releases/) by [this](https://github.com/django/django/commit/2d2c1d0c97832860fbd6597977e2aae17dd7e5b2) commit.",
    "severity": "MEDIUM",
    "cvss2": {
        "baseScore": 5.1,
        "impactSubscore": 6.4,
        "exploitabilitySubscore": 4.9,
        "severity": "LOW",
        "accessVector": "NETWORK",
        "accessComplexity": "HIGH",
        "authentication": "NONE",
        "confidentialityImpact": "PARTIAL",
        "integrityImpact": "PARTIAL",
        "availabilityImpact": "PARTIAL",
        "temporalMetrics": {
            "exploitability": "UNPROVEN",
            "remediationLevel": "OFFICIAL_FIX",
            "reportConfidence": "CONFIRMED",
            "score": 3.8
        },
        "vector": "(AV:N/AC:H/Au:N/C:P/I:P/A:P/E:U/RL:OF/RC:C)"
    },
    "cvss3": {
        "baseScore": 5.6,
        "impactSubscore": 3.4,
        "exploitabilitySubscore": 2.2,
        "severity": "MEDIUM",
        "attackVector": "NETWORK",
        "attackComplexity": "HIGH",
        "confidentialityImpact": "LOW",
        "integrityImpact": "LOW",
        "availabilityImpact": "LOW",
        "privilegesRequired": "NONE",
        "scope": "UNCHANGED",
        "userInteraction": "NONE",
        "temporalMetrics": {
            "exploitability": "UNPROVEN",
            "remediationLevel": "OFFICIAL_FIX",
            "reportConfidence": "CONFIRMED",
            "score": 4.9
        },
        "vector": "CVSS:3.1/AV:N/AC:H/PR:N/UI:N/S:U/C:L/I:L/A:L/E:U/RL:O/RC:C"
    },
    "useCvss3": true,
    "classifications": [],
    "zeroDay": false,
    "underReview": false,
    "parentAdvisory": false,
    "workaround": "",
    "vendorFixDate": "2021-05-06T00:00:00.000Z",
    "bdsaTags": [],
    "overallScore": 4.9,
    "_meta": {
        "allow": [
            "GET"
        ],
        "href": "https://trendmicro2.app.blackduck.com/api/vulnerabilities/BDSA-2021-1271",
        "links": [
            {
                "rel": "cwes",
                "href": "https://trendmicro2.app.blackduck.com/api/cwes/CWE-93"
            },
            {
                "rel": "related-vulnerability",
                "href": "https://trendmicro2.app.blackduck.com/api/vulnerabilities/CVE-2021-32052",
                "label": "NVD"
            },
            {
                "rel": "bdsa-ranges",
                "href": "https://trendmicro2.app.blackduck.com/api/vulnerabilities/BDSA-2021-1271/ranges"
            },
            {
                "rel": "reference",
                "href": "https://bugs.python.org/issue43882",
                "label": "LINK"
            },
            {
                "rel": "reference",
                "href": "https://github.com/django/django/commit/2d2c1d0c97832860fbd6597977e2aae17dd7e5b2",
                "label": "PATCH"
            },
            {
                "rel": "reference",
                "href": "https://github.com/django/django/commit/afb23f5929944a407e4990edef1c7806a94c9879",
                "label": "PATCH"
            },
            {
                "rel": "reference",
                "href": "https://github.com/django/django/commit/d9594c4ea57b6309d93879805302cec9ae9f23ff",
                "label": "PATCH"
            },
            {
                "rel": "reference",
                "href": "https://www.djangoproject.com/weblog/2021/may/06/security-releases/",
                "label": "ADVISORY"
            },
            {
                "rel": "reference",
                "href": "https://www.djangoproject.com/weblog/2021/may/06/security-releases/",
                "label": "VENDOR_UPGRADE"
            },
            {
                "rel": "default-remediation-status",
                "href": "https://trendmicro2.app.blackduck.com/api/vulnerabilities/BDSA-2021-1271/default-remediation-status"
            }
        ]
    }
}